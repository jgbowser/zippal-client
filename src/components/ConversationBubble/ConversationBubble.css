.ConversationBubble__convo_card {
  position: relative;
}

.small {
  transition: all 400ms;
  margin: 4px 4px;
  width: 300px;
  height: 300px;
  /* background-color: lightgreen; */
  background-image: linear-gradient(360deg, white, darkkhaki);
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
  text-align: center;
}

.small:hover {
  width: 315px;
  height: 315px;
}

.expanded {
  z-index: 3;
  height: 55vh;
  width: 75vw;
  background-color: lightgray;
  position: absolute;
  padding: 20px;
  border-radius: 5px;
  border: 1px solid black;
  transition: all 0.25s cubic-bezier(0, 0.56, 0.69, 0.49); /*make it smooth*/
  z-index: 3;
}

.expanded.show {
  visibility: visible;
  z-index: 200;
  opacity: 1;
  top: 50%; /*slide it down smoothly*/
}

.expanded.show ~ .overlay {
  opacity: 1;
  visibility: visible;
  z-index: 100;
}
