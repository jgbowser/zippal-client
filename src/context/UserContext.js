import React, {Component} from 'react'
import AuthApiService from '../services/auth-api-service'
import UserService from '../services/user-api-service'
import TokenService from '../services/token-service'
import IdleService from '../services/idle-service'

const UserContext = React.createContext({
  user:{},
  profileInfo:{},
  error:null,
  setError:() => { },
  clearError: () => { },
  setUser: () => { },
  getProfile: () => { },
  updateProfile: () => { },
  processLogin: () => { },
  processLogout: () => { },
})

export default UserContext

export class UserProvider extends Component {
  constructor(props){
    super(props)
    const state = {user: {}, error:null, profileInfo: {}}

    const jwtPayload = TokenService.parseAuthToken()
    

    if(jwtPayload){
      const profileData = this.getProfile()

      state.user = {
        id: jwtPayload.id,
        display_name: jwtPayload.display_name,
        username: jwtPayload.sub,
      }

      if(profileData)
        state.profileInfo = {
          bio: profileData.bio,
          display_name: profileData.display_name,
          fa_icon: profileData.fa_icon,
          location: profileData.location
      }
    }
      

    this.state = state;
    IdleService.setIdleCallback(this.logoutBecauseIdle)
  }

  componentDidMount(){
    if(TokenService.hasAuthToken()){
      IdleService.registerIdleTimerResets()
      TokenService.queueCallbackBeforeExpiry(() => {
        this.fetchRefreshToken()
      })
    }
  }

  componentWillUnmount() {
    IdleService.unRegisterIdleResets()
    TokenService.clearCallBackBeforeExpiry()
  }

  setError = error => {
    console.error(error)
    this.setState({error})
  }

  clearError = () => {
    this.setState({error: null})
  }
  
  setUser = user => {
    this.setState({user})
  }

  getProfile = async () => {
    const profileInfo = await UserService.getUserProfile()
    this.setState({profileInfo})
  }

  updateProfile = (profileInfo) => {
    this.setState({profileInfo})
  }

  processLogin = authToken => {
    TokenService.saveAuthToken(authToken)

    const jwtPayload = TokenService.parseAuthToken()
    this.setUser({
      id: jwtPayload.id,
      display_name: jwtPayload.display_name,
      username: jwtPayload.sub,
    })
    this.getProfile()
    IdleService.registerIdleTimerResets()
    TokenService.queueCallbackBeforeExpiry(() => {
      this.fetchRefreshToken()
    })
  }

  processLogout = () => {
    TokenService.clearAuthToken()
    TokenService.clearCallBackBeforeExpiry()
    IdleService.unRegisterIdleResets()
    this.setUser({})
    this.setState({profileInfo: {}})
  }

  logoutBecauseIdle = () => {
    TokenService.clearAuthToken()
    TokenService.clearCallBackBeforeExpiry()
    IdleService.unRegisterIdleResets()
    this.setUser({idle:true})
  }

  fetchRefreshToken = () => {
    AuthApiService.refreshToken()
      .then(res => {
        TokenService.saveAuthToken(res.authToken)
        TokenService.queueCallbackBeforeExpiry(() => {
          this.fetchRefreshToken()
        })
      })
      .catch(err => {
        this.setError(err)
      })
  }

  render() {
    const value = {
      user: this.state.user,
      profileInfo: this.state.profileInfo,
      error: this.state.error,
      setError: this.setError,
      clearError: this.clearError,
      setUser: this.setUser,
      getProfile: this.getProfile,
      updateProfile: this.updateProfile,
      processLogin: this.processLogin,
      processLogout: this.processLogout
    }
    return (
      <UserContext.Provider value={value}>
        {this.props.children}
      </UserContext.Provider>
    )
  }
}