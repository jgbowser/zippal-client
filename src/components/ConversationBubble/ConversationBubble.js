import React, {Component} from 'react';
import UserContext from '../../context/UserContext'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import Message from '../Message/Message'
import ConversationNotification from '../ConversationNotification/CoversationNotification'
import'./ConversationBubble.css'
import MessageApiService from '../../services/message-api-service';

export default class ConversationBubble extends Component {
  static contextType = UserContext 
  
  constructor(props) {
    super(props)
    this.state = {
      expanded: false,
      selectedMessage: null,
      newMessage: null,
      confirmEndConvoPanel: false
    }
  }
  
  toggleBubble = () => {
    this.setState({
      expanded: !this.state.expanded
    })
  }

  formatDate = (date) => {
    return new Date(date).toDateString()
  }


  
  messageStatusMessage = () => {
    const { messageData } = this.props
    if(!messageData) {
      return 'No messages in this conversation'
    }
    if(messageData.length === 0) {
      return 'No messages in this conversation'
    }
    const lastMessage = messageData[messageData.length - 1]
    //if logged in user was the sender of the most recent message
    if(lastMessage.sender_id === this.context.user.id) {
      //message has been created but not sent
      if(lastMessage.sender_status === 'Pending') {
        return 'Keep working on your current draft'
        //message has been sent, other user hasn't started new reply yet
      } else if(lastMessage.sender_status === 'Sent'){
        return 'Your message is on it\'s way'
      } else if(lastMessage.sender_status === 'Awaiting Reply') {
        return 'Your pal has seen your message'
      }
      //if logged in user is the receiver of the most recent message
    } else if(lastMessage.sender_id !== this.context.user.id) {
      // Pal has seen logged in user's message and started a reply
        if(lastMessage.sender_status === 'Pending' && !lastMessage.is_read) {
        return 'Your pal is working on their response'
      } else if(lastMessage.sender_status === 'Sent') {
        return 'You have a new message(or on the way once we get the timeout)'
      } else if(lastMessage.sender_status === 'Awaiting Reply') {
        return 'Start a reply message'
      }
    }
  }

  renderSmallView = () => {
    return (
      <div className='ConversationBubble__convo_card small' >
        {this.props.messageData.length !== 0 ? <ConversationNotification messageData={this.props.messageData} /> : ''}
        <h2>{this.props.convoData.pal_name}</h2>
        <button onClick={this.toggleBubble}><FontAwesomeIcon className='ConversationBubble__pal_icon' icon={this.props.convoData.fa_icon} /></button>
        <p>Conversation Status: {this.messageStatusMessage()}</p>
        <p>Total Messages: {this.props.messageData.length}</p>
      </div>
    )
  }

  //create messsage containers that display sender name and date message was sent
  //these messages can be selected to view the content
  renderMessages = () => {
    const { messageData } = this.props
    
    if(messageData.length === 0) {
      return <span>No messages to display yet</span>
    }

    const messageContainers = messageData.map((message, i) => {
      return (
        <button onClick={() => this.selectMessageHandler(i)} key={message.id} className='ConersationBubble__message_select'>
          <p>{message.sender_id === this.context.user.id ? 'Outgoing': 'Incoming'}</p>
          <p>Date sent: {this.formatDate(message.date_sent)}</p>
          <p>Content: {message.content.substring(0, 30)}...</p>
        </button>
      )
    })

    return messageContainers
  }

  selectMessageHandler = (id) => {
    const selected = this.props.messageData[id]
    this.setState({
      selectedMessage: selected,
    })
    if (!selected.is_read && selected.receiver_id === this.context.user.id) {
        MessageApiService.readMessage(selected)
      }
  }

  clearSelectedMessage = () => {
    this.setState({
      selectedMessage: null
    })
  }

  // confirm user actually wants to delete conversation
  confirmEndConvo = () => {
    this.setState({
      confirmEndConvoPanel: true
    })
  }

  // cancel the end conversation panel
  cancelEndConvo = () => {
    this.setState({
      confirmEndConvoPanel: false
    })
  }

  // render the confirm delete conversation panel
  renderConfirmEndConvoPanel = () => {
    const {handleEndConvo} = this.props
    let convo = (this.props.convoData)

    return (
      <>
      <p>Are you sure you want to delete this conversation? You will lose all of your letters and this cannot be undone.</p>
      <button className="ConversationBubble__end_convo_confirm_btn" onClick={() => handleEndConvo(convo)}>Confirm</button>
      <button className="ConversationBubble__end_convo_confirm_btn" onClick={this.cancelEndConvo}>Cancel</button>
      </>
    )
  }

  // conditionally render reply(create new message) button or continue draft(open last message in text area to continue writing)
  renderExpandedView = () => {
    return (
    <div className='ConversationBubble__convo_card expanded'>
      {this.state.selectedMessage ? <button onClick={this.clearSelectedMessage}>Go back</button>: ''}
      <button onClick={this.toggleBubble}><FontAwesomeIcon className='ConversationBubble__pal_icon' icon={this.props.convoData.fa_icon} /></button>
      <button className="ConversationBubble__end_convo_btn" onClick={this.confirmEndConvo}>
        End Conversation
      </button>
      {!this.state.selectedMessage ? this.renderMessages() : ''}
      {this.state.selectedMessage ? <Message convoData={this.props.convoData} newMessage={this.state.selectedMessage} setNewMessage={this.props.setNewMessage} clearSelectedMessage={this.clearSelectedMessage}/>: ''}
      {this.state.confirmEndConvoPanel ? this.renderConfirmEndConvoPanel() : ''}
    </div>
    )
  }




  
  render() {
    const { expanded } = this.state
    return (
      <>
        {expanded === false ? this.renderSmallView() : this.renderExpandedView()}
      </>
    )
  }
}