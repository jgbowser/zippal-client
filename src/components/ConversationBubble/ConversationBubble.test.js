import React from 'react'
import ReactDOM from 'react-dom'
import testHelpers from '../../services/test-helpers'
import ConversationBubble from './ConversationBubble'
import UserContext from '../../context/UserContext'

describe('ConversationBubble Component', () => {
  const testConvos = testHelpers.makeTestConvos()
  const testConvo = testConvos[0]
  const testMessages = testHelpers.makeTestMessages()
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
    <UserContext.Provider value={{user: {id: 1}}}>
      <ConversationBubble convoData={testConvo} messageData={testMessages}/>
    </UserContext.Provider>
    , div)
    ReactDOM.unmountComponentAtNode(div)
  })
})