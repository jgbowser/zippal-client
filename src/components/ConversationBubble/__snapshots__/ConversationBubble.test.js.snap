// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConversationBubble Component renders ConversationBubble component 1`] = `
<div
  className="ConversationBubble__convo_card small"
>
  <div
    className="ConversationNotification__icon_container blue"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-pen-nib fa-w-16 ConversationNotification__icon"
      data-icon="pen-nib"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M136.6 138.79a64.003 64.003 0 0 0-43.31 41.35L0 460l14.69 14.69L164.8 324.58c-2.99-6.26-4.8-13.18-4.8-20.58 0-26.51 21.49-48 48-48s48 21.49 48 48-21.49 48-48 48c-7.4 0-14.32-1.81-20.58-4.8L37.31 497.31 52 512l279.86-93.29a64.003 64.003 0 0 0 41.35-43.31L416 224 288 96l-151.4 42.79zm361.34-64.62l-60.11-60.11c-18.75-18.75-49.16-18.75-67.91 0l-56.55 56.55 128.02 128.02 56.55-56.55c18.75-18.75 18.75-49.15 0-67.91z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <h2>
    Test User
  </h2>
  <button
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-user-circle fa-w-16 ConversationBubble__pal_icon"
      data-icon="user-circle"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 496 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
  <p>
    Conversation Status: 
    Keep working on your current draft
  </p>
  <p>
    Total Messages: 
    1
  </p>
</div>
`;
